<h1 style="text-align: center; color: purple;" markdown="1">Econ220 Python Lab: Pandas Data 
Manipulation</h1>
<h2 style="text-align: center; color: #012169" markdown="1">Handout #5</h2>

### Table of Contents
* [Pandas for Data Manipulation](#anchor1)
    * [The Package Set Up](#anchor2)
* [Data Manipulation with Pandas](#anchor3)
    * [pandas functions pd.](#anchor4) 
    * [Using `[ ]` to Select Observations](#anchor5)
    * [Your Turn!](#practice)
    * [Using `.sort_values()` to Order Observations](#anchor6)
    * [Using `.agg()` for Summary Statistics](#anchor7)
    * [Finding Statistics by Group using `.groupby()`](#anchor8)
    * [In Class Exercise](#anchor9)




# Pandas for Data Manipulation <a class = anchor id = anchor1 ></a>


#### The Package Set Up<a class = anchor id = anchor2 ></a>

In [58]:
# (Student)import package os to find the path of your working directory
import os
path = os.getcwd()
print(path)
# (Student)clarify if necessary why the path doesn't have your name in it 

# Import other necesarry packages 
import numpy as np
import pandas as pd


/Users/gavinmason


In this handout we will simply load a fun data set about baby names. The file is on CSV, let's load (import) it into python and inspect the data by using some methods from pandas. 
* `.head()` provides the first 5 observations by default
* `.info()` provides the structure of the data
* `.describe()`a summary of each variable in the data set showing the main descriptive statistics


In [59]:
# Read CSV file into a DataFrame
babynames = pd.read_csv('https://www.alexandramanta.info/media/econ220_labs/' + 'NationalNames.csv')

In [60]:
# Autograder cell -do not delete it

In [61]:
# the file has a variable called Id that we do not need I will use the function del to get rid of it in the dataframe
del babynames["Id"]
babynames.head()

Name  Year Gender  Count
0       Mary  1880      F   7065
1       Anna  1880      F   2604
2       Emma  1880      F   2003
3  Elizabeth  1880      F   1939
4     Minnie  1880      F   1746

In [62]:
# Autograder cell -do not delete it

In [63]:
# . info() provides the structure of the data
babynames.info()
# . describe() provides you with a good summary of each variable in the data set showing the main descriptive statistics
babynames.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1825433 entries, 0 to 1825432
Data columns (total 4 columns):
 #   Column  Dtype 
---  ------  ----- 
 0   Name    object
 1   Year    int64 
 2   Gender  object
 3   Count   int64 
dtypes: int64(2), object(2)
memory usage: 55.7+ MB


Year         Count
count 1825433.00000 1825433.00000
mean     1972.61978     184.68792
std        33.52891    1566.71098
min      1880.00000       5.00000
25%      1949.00000       7.00000
50%      1982.00000      12.00000
75%      2001.00000      32.00000
max      2014.00000   99680.00000

The data we loaded is from the babynames dataset which contains information about the number of children (of either sex) who were given a certain name from 1880 to 2015. All names used 5 or more times are included. As you can see from the summary, the data set contains five variables: year, sex, name, n the number of times the name was used, prop de proportion of times the name was used.  

# Data Manipulation with `Pandas` <a class = anchor id = anchor3 ></a>

## `pandas` methods `pd.` <a class = anchor id = anchor4 ></a>

First, in this class I might use the word method and function interchangeable. A method is a function that belongs to and object. It is ussually writen as **`object.method()`**

Let **df** be the name of a given data frame ( in this handout we will later replace df by babynames)

function |  Purpose
-----------|--------
df[df["column_name"] + comparision sign +  criteria]  | Select rows based on criteria
 df.sort_values()  | order rows
 df[ [column_name] ]  | choose columns
 df.drop_duplicates() | unique values of column(s)
 df ["new_column"] = criteria | Create new columns
 df.agg()  | Find a summary statistic
 df.sample( n ) | Take a random sample of size n
 df.groupby() | Creates groups based on the values or categories of a variable


# Filtering observations 

Remember that the observations are represented by the rows. Filtering the observations means that you will make a selection of a set of observations or rows according to a criteria. This is like taking a horizontal slice of your data. 

See code below

>`df[expression]` for example `df[ df['var'] == value & df['var'] == value ]

**Filter by gender and year**

In [64]:
# We want to restrict the rows to those of female names in the year 1997 and show the first 10 observations
# Follow by.head() so that we can show how the data looks and we .reset_index() so it just shows an index from 0 to 10 
babynames[(babynames["Gender"]=="F")&(babynames["Year"]==1997)].head(10).reset_index()

index       Name  Year Gender  Count
0  1249138      Emily  1997      F  25731
1  1249139    Jessica  1997      F  21043
2  1249140     Ashley  1997      F  20894
3  1249141      Sarah  1997      F  20690
4  1249142     Hannah  1997      F  20586
5  1249143   Samantha  1997      F  20169
6  1249144     Taylor  1997      F  19503
7  1249145     Alexis  1997      F  17172
8  1249146  Elizabeth  1997      F  15411
9  1249147    Madison  1997      F  15187

Roughly translates to: Take the babynames dataset **then** keep only those observations that are female names and in 1997 **then** show the first 10 observations of that data set. If you execute this code, the output, shown above, will be the first 10 observations of a dataset that includes only female names given in born in 1997.
 
### Using `[ ]` to Filter Observations <a class = anchor id = anchor5 ></a>

Let's practice what you learned. 
* Create a new dataset (object) with just boys, name it just_boys. 
* Create a dataset with the names given to babies since 1995
* Create a dataset of just girls born between 1999 and 2005, name this dataset girls99_05



In [65]:
# one object 
# create the boolean expression first 
# A boolean expression returns true or false 
is_male = babynames["Gender"]=="M"
print(is_male.head(3))

0    False
1    False
2    False
Name: Gender, dtype: bool


In [66]:
# Autograder cell -do not delete it

In [67]:
#Subsetting/filtering dataframe
# Using the boolean expression
just_boys = babynames[is_male]

# The boolean expression directly inside df[] to filter
just_boys = babynames[(babynames["Gender"]=="M")]
just_boys = babynames[(babynames["Gender"]!="F")]

In [68]:
# Autograder cell -do not delete it

In [69]:
# Many expressions in one 
after_1995 = babynames[(babynames["Year"]>=1995)]
girls99_05 = babynames[(babynames["Gender"]=="F") & (babynames["Year"]>=1999) & (babynames["Year"]<=2005)]
girls99_05.head(6).reset_index() # looks at the first 6 rows 

index      Name  Year Gender  Count
0  1303994     Emily  1999      F  26537
1  1303995    Hannah  1999      F  21667
2  1303996    Alexis  1999      F  19232
3  1303997     Sarah  1999      F  19084
4  1303998  Samantha  1999      F  19033
5  1303999    Ashley  1999      F  18131

In [70]:
# Autograder cell -do not delete it

### Using `.sort_values()` to Order Observations <a class = anchor id = anchor6 ></a>

You can order the observations by columns, lets check 

In [71]:
#Sort names by year in descending order 
bsorted1 = babynames.sort_values(["Year", "Count"], ascending=(True, False))
print(bsorted1.head(3))
bsorted1.tail(3)
#include tail and also sorting by a couple of bvarianl 

        Name  Year Gender  Count
942     John  1880      M   9655
943  William  1880      M   9532
0       Mary  1880      F   7065


Name  Year Gender  Count
1825430  Zymiere  2014      M      5
1825431    Zyran  2014      M      5
1825432    Zyrin  2014      M      5

In [72]:
# Autograder cell -do not delete it

### Your turn: Filter and nlargest. <a class = anchor id = practice ></a>
Find the 6 most popular names for your gender in the year you where born. Below are two ways of doing it,

* `sort_values()`
* `nlargest(n, 'varname', keep='all')`

In [73]:
 babynames[(babynames["Gender"]=="M")&(babynames["Year"]==1999)].sort_values(["Count"], ascending=[False]).head(6)

Name  Year Gender  Count
1320932        Jacob  1999      M  35344
1320933      Michael  1999      M  33905
1320934      Matthew  1999      M  30414
1320935       Joshua  1999      M  27251
1320936     Nicholas  1999      M  25635
1320937  Christopher  1999      M  25601

In [74]:
 babynames[(babynames["Gender"]=="F")&(babynames["Year"]==2000)].nlargest(6,"Count",keep="all")

Name  Year Gender  Count
1332538    Emily  2000      F  25952
1332539   Hannah  2000      F  23073
1332540  Madison  2000      F  19967
1332541   Ashley  2000      F  17995
1332542    Sarah  2000      F  17687
1332543   Alexis  2000      F  17627

### Other examples


Suppose hypothetically you wanted to create a new variable called "popular" when the names where chosen more that 10 thousand times. To create a new dataset that included this variable

`df['newvarname'] =  The expression or function that creates the new variable`

In [75]:
#Adding a new column variable to a dataframe
babynames["popular"] = babynames["Count"]>1000
babynames.info()
babynames.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1825433 entries, 0 to 1825432
Data columns (total 5 columns):
 #   Column   Dtype 
---  ------   ----- 
 0   Name     object
 1   Year     int64 
 2   Gender   object
 3   Count    int64 
 4   popular  bool  
dtypes: bool(1), int64(2), object(2)
memory usage: 57.4+ MB


Name  Year Gender  Count  popular
0       Mary  1880      F   7065     True
1       Anna  1880      F   2604     True
2       Emma  1880      F   2003     True
3  Elizabeth  1880      F   1939     True
4     Minnie  1880      F   1746     True

In [76]:
# Autograder cell -do not delete it

### Selecting columns 


* Use `[[  ]]` to select columns from the dataframe and maintain the dataframe structure
* `drop([var list], axis=1)` can also work
* df.loc['rowmame' or 'colname'] allows you to access variables by names too as we saw in our previous handout

In [77]:
#Selecting columns using their labels
no_popular = babynames[["Name","Year","Gender"]]
no_popular.head()

Name  Year Gender
0       Mary  1880      F
1       Anna  1880      F
2       Emma  1880      F
3  Elizabeth  1880      F
4     Minnie  1880      F

In [78]:
babynames.drop(["Count","popular"], axis=1).head()

Name  Year Gender
0       Mary  1880      F
1       Anna  1880      F
2       Emma  1880      F
3  Elizabeth  1880      F
4     Minnie  1880      F

In [79]:
# Autograder cell -do not delete it

We could also use `.iloc` or `.loc` as seen before
* df.loc['rowmame', 'colname'] access observations and variables by names 
* df.iloc[i, j] access observations and variables by positions i and j (remember indices start at 0 in Python)

### Using numpy methods and `agg()` for Summary Statistics <a class = anchor id = anchor7 ></a>

Suppose we wanted to find the average popularity of names, the number of observations in the data set, and maximum popularity in the same year. 

**Using numpy we can apply methods to the flatten array or column see below**

In [80]:
pd.set_option('display.float_format', lambda x: '%.5f' % x)# remove scientific notation
print('Mean:')
print(babynames["Count"].mean())
print(f'Sum:')
print(babynames["Count"].sum())
print(f'Max:')
print(babynames["Count"].max())
print(f'Min:')
print(babynames["Count"].min())
print(f'Count:')
print(babynames["Count"].count())

Mean:
184.68792116719703
Sum:
337135426
Max:
99680
Min:
5
Count:
1825433


**Using .agg() from pandas we can apply methods to the column or columns see below**

In [81]:
babynames["Count"].agg(["mean","sum","count","max","size"])

mean          184.68792
sum     337135426.00000
count     1825433.00000
max         99680.00000
size      1825433.00000
Name: Count, dtype: float64

In [82]:
# You need two brackets [[ ]] because by choosing two series you are dealing with a dataframe
babynames[["Count","popular"]].agg(["max","mean"])

Count popular
max  99680.00000    True
mean   184.68792 0.02764

### Finding Statistics by Group using `.groupby()` <a class = anchor id = anchor8 ></a>

The `.groupby()` functions divides the data into groups for all future commands.  This can be extremely useful in combination with the `agg()` method.  For example, suppose we wanted to find the average popularity of a name by sex in all this time (**name your dataset here "gr_babynames" - this is not in the video but please do so for autograder to grade your code answer**):

In [83]:
# for count also consider 'size'. size will return n for NaN values also, whereas 'count' will not.
babynames.groupby(["Year","Gender"])["Count"].agg(["mean","sum","count"]).unstack('Gender') 

mean                sum           count       
Gender         F         M        F        M      F      M
Year                                                      
1880    96.59554 104.43384    90993   110491    942   1058
1881    98.03198 101.04814    91954   100745    938    997
1882   104.91245 103.44677   107850   113688   1028   1099
1883   106.56641 101.58155   112321   104629   1054   1030
1884   110.08703 101.72889   129022   114445   1172   1125
...          ...       ...      ...      ...    ...    ...
2010    89.53222 134.39021  1772738  1913851  19800  14241
2011    89.73900 132.12576  1753500  1893230  19540  14329
2012    90.09256 132.90757  1753922  1889414  19468  14216
2013    90.94570 134.27512  1745339  1881463  19191  14012
2014    92.76630 136.03606  1768775  1901376  19067  13977

[135 rows x 6 columns]

In [84]:
# Autograder cell -do not delete it

Note that the `.unstack(<var2>)` in this case Gender, here is purely cosmetic: it transforms the data from "long" to "wide" which is how crosstabs are usually presented. Try and see what happens if you remove it



                     

# In Class Exercise <a class = anchor id = anchor9></a>

Find the six names that were most popular female and male names in the **year 1999**. 

In this case the steps are 

* Filter data for the year
* Sort the data by columns gender and count, in descending order
* Group by gender to show both genders
* use .head(6) to show the top 6 and 
* reset index so the indexing looks nice 
* **Name your dataset "df"**.

The order of the steps (chaining methods) is not obvious, this takes some practice and trial and error. **Be patient with yourself**


In [85]:
 df = babynames[(babynames["Year"]==2001)].sort_values(["Count","Gender"], ascending=[False, False])
df.head(6).reset_index()

index         Name  Year Gender  Count  popular
0  1380267        Jacob  2001      M  32527     True
1  1380268      Michael  2001      M  29675     True
2  1380269      Matthew  2001      M  26799     True
3  1380270       Joshua  2001      M  26006     True
4  1362301        Emily  2001      F  25052     True
5  1380271  Christopher  2001      M  23117     True

In [86]:
# Autograder cell -do not delete it

&nbsp;

<hr />

<p style="text-align: center;">ECON220 R Programing Laboratory</a></p>
<p style="text-align: center;">Professor <em> Paloma Lopez de mesa Moyano</em></a></p>
<p style="text-align: center;"><span style="color: #6666FF;"><em>paloma.moyano@emory.edu</em></span></p>

<p style="text-align: center;">Department of Economics</a></p>
<p style="text-align: center;">Emory University</a></p>

&nbsp;
    
    

In [16]:
#!jupyter nbconvert --to html Econ220Lab_class4_PandasManipulation.ipynb